extension/src/state_aggregate.rs

```diff
@@ -0,0 +1,618 @@
+                None,
+            );
+            let result: i64 = client
+                .select(
+                    r#"
+SET parallel_setup_cost = 0;
+SET parallel_tuple_cost = 0;
+SET min_parallel_table_scan_size = 0;
+SET max_parallel_workers_per_gather = 4;
+SET parallel_leader_participation = off;
+SET enable_indexonlyscan = off;
+SELECT toolkit_experimental.duration_in('one', toolkit_experimental.state_agg(ts, state)) FROM (
+    SELECT * FROM test
+    UNION ALL SELECT * FROM test
+    UNION ALL SELECT * FROM test
+    UNION ALL SELECT * FROM test) u
+                "#,
+                    None,
+                    None,
+                )
+                .first()
+                .get_one()
+                .unwrap();
+            assert_eq!(172800000000, result);
+        });
+        use std::sync::atomic::Ordering::Relaxed;
+        assert!(counters::COMBINE_NONE.load(Relaxed) == 0); // TODO untested
+        assert!(counters::COMBINE_A.load(Relaxed) == 0); // TODO untested
+        assert!(counters::COMBINE_B.load(Relaxed) > 0); // tested
+        assert!(counters::COMBINE_BOTH.load(Relaxed) > 0); // tested
+    }
+
+    // the sample query from the ticket
+    #[pg_test]
+    fn sample_query() {
+        Spi::execute(|client| {
+            client.select("CREATE TABLE test(ts timestamptz, state TEXT)", None, None);
+            client.select(
+                r#"INSERT INTO test VALUES
+                    ('2020-01-01 00:00:00+00', 'START'),
+                    ('2020-01-01 00:01:00+00', 'ERROR'),
+                    ('2020-01-01 00:02:00+00', 'STOPPED')"#,
+                None,
+                None,
+            );
+            assert_eq!(
+                client
+                    .select(
+                        r#"SELECT toolkit_experimental.duration_in('ERROR', states) as error,
+                                  toolkit_experimental.duration_in('START', states) as start,
+                                  toolkit_experimental.duration_in('STOPPED', states) as stopped
+                             FROM (SELECT toolkit_experimental.state_agg(ts, state) as states FROM test) as foo"#,
+                        None,
+                        None,
+                    )
+                    .first()
+                    .get_three::<i64, i64, i64>(),
+                (Some(60000000), Some(60000000), Some(0))
+            );
+        })
+    }
+
+    #[pg_test]
+    #[should_panic(expected = "state cannot be both ERROR and START at 631152000000000")]
+    fn two_states_at_one_time() {
+        Spi::execute(|client| {
+            client.select("CREATE TABLE test(ts timestamptz, state TEXT)", None, None);
+            client.select(
+                r#"INSERT INTO test VALUES
+                        ('2020-01-01 00:00:00+00', 'START'),
+                        ('2020-01-01 00:00:00+00', 'ERROR')"#,
+                None,
+                None,
+            );
+            // TODO why isn't should_panic working?
```

Comment:
This doesn't need to block merge either; I can just remove the test for now.

If we expect to report errors to the user, we ought to test them, so I don't want to lose this forever.